app/src/main/java/com/jalasoft/routesapp/ui/auth/phoneAuthentication/view/PhoneAuthenticationFragment.kt

```diff
@@ -0,0 +1,173 @@
+            viewModel.sendVerificationCode(countryCode, phoneNumber, requireActivity())
+        }
+    }
+    private fun observers() {
+        val errorObserver = Observer<String> { errorMessage ->
+            binding.tvError.isVisible = true
+            binding.tvError.text = errorMessage
+        }
+        val alertDialogErrorObserver = Observer<String> { errorMessage ->
+            textInputLayout.error = errorMessage
+        }
+        val codeSubmittedObserver = Observer<Boolean> {
+            if (dialogInitialized) {
+                textInputLayout.error = RoutesAppApplication.resource?.getString(R.string.reg_val_phone_code_forwarded).toString()
+            } else {
+                initAlertDialog()
+            }
+        }
+        val onAutomaticVerificationObserver = Observer<Boolean> {
+            viewModel.verifyCode("", requireActivity())
+        }
+        val onVerificationCompletedObserver = Observer<Boolean> {
+            dialog.dismiss()
+            findNavController().navigate(R.id.homeFragment)
+        }
+        viewModel.errorMessage.observe(this, errorObserver)
+        viewModel.alertDialogErrorMessage.observe(this, alertDialogErrorObserver)
+        viewModel.codeSubmitted.observe(this, codeSubmittedObserver)
+        viewModel.onAutomaticVerification.observe(this, onAutomaticVerificationObserver)
+        viewModel.onVerificationCompleted.observe(this, onVerificationCompletedObserver)
+    }
+
+    private fun getEditTextLayout(context: Context): ConstraintLayout {
+        val constraintLayout = ConstraintLayout(context)
+        val layoutParams = ConstraintLayout.LayoutParams(
+            ConstraintLayout.LayoutParams.MATCH_PARENT,
+            ConstraintLayout.LayoutParams.WRAP_CONTENT
+        )
+        constraintLayout.layoutParams = layoutParams
+        constraintLayout.id = View.generateViewId()
+
+        val textInputLayout = TextInputLayout(context)
+        textInputLayout.boxBackgroundMode = TextInputLayout.BOX_BACKGROUND_OUTLINE
+        layoutParams.setMargins(
+            32.toDp(context),
+            8.toDp(context),
+            32.toDp(context),
+            8.toDp(context)
+        )
+        textInputLayout.layoutParams = layoutParams
+        textInputLayout.hint = RoutesAppApplication.resource?.getString(R.string.reg_val_code).toString()
+        textInputLayout.id = View.generateViewId()
+        textInputLayout.tag = "textInputLayout"
+
+        val codeEditText = TextInputEditText(context)
+        codeEditText.id = View.generateViewId()
+        codeEditText.tag = "codeEditText"
+        codeEditText.inputType = InputType.TYPE_CLASS_NUMBER
+
+        textInputLayout.addView(codeEditText)
+
+        val constraintSet = ConstraintSet()
+        constraintSet.clone(constraintLayout)
+
+        constraintLayout.addView(textInputLayout)
+        return constraintLayout
+    }
+
+    private fun Int.toDp(context: Context): Int = TypedValue.applyDimension(
```

Comment:
if this is an Utility to convert from pixels to DP, it should be moved to its own utilities file

Reply:
done